### PR TITLE
Create/Update tenant with ReCAPTCHA Config

### PR DESCRIPTION
**Pull request by Xiaoshouzi-gh**
Support reCaptcha config /create update on tenants.
     - Updated reCaptcha config type align with backend API.
     - Support create and update tenants with reCaptcha config.
     - Added reCaptcha unit tests on tenants.

### Discussion

  * Read the contribution guidelines (CONTRIBUTING.md).
  * If this has been discussed in an issue, make sure to link to the issue here. 
    If not, go file an issue about this **before creating a pull request** to discuss.

### Testing

  * Make sure all existing tests in the repository pass after your change.
  * If you fixed a bug or added a feature, add a new test to cover your code.

### API Changes

  * At this time we cannot accept changes that affect the public API.  If you'd like to help 
    us make Firebase APIs better, please propose your change in an issue so that we 
    can discuss it together.

### CODE REVIEW

File: src/auth/tenant.ts [hunk 1]
```diff
@@ -21,7 +21,7 @@ import { AuthClientErrorCode, FirebaseAuthError } from '../utils/error';
 import {
   EmailSignInConfig, EmailSignInConfigServerRequest, MultiFactorAuthServerConfig,
   MultiFactorConfig, validateTestPhoneNumbers, EmailSignInProviderConfig,
-  MultiFactorAuthConfig,
+  MultiFactorAuthConfig, RecaptchaConfigAuth, RecaptchaConfig
```
Comment (renkelvin) [inline src/auth/tenant.ts:24]: Change to *AuthConfig to match MultiFactorAuthConfig?

File: test/unit/auth/tenant.spec.ts [hunk 2]
```diff
@@ -79,6 +79,55 @@ describe('Tenant', () => {
     },
   };
 
+  const serverRequestWithRecaptcha: TenantServerResponse = {
```
Comment (renkelvin) [inline test/unit/auth/tenant.spec.ts:82]: serverResponse?

**Review by renkelvin — COMMENTED** [review]
Looks good to me. Please add Lahiru for approval.

File: src/auth/tenant.ts (hunk 1 above)
Comment (Xiaoshouzi-gh) [reply src/auth/tenant.ts]: Done.

File: test/unit/auth/tenant.spec.ts (hunk 2 above)
Comment (Xiaoshouzi-gh) [reply test/unit/auth/tenant.spec.ts]: Done

**Review by renkelvin — APPROVED** [review]

**Review by lahirumaramba — CHANGES_REQUESTED** [review]
Thank you, Liubin!
I added some comments and questions. Didn't get a chance to dive into unit tests. I will take another look later. Thanks!

File: src/auth/auth-config.ts [hunk 3]
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
   /**
    * The reCAPTCHA site key.
    */
-   key: string;
+  key: string;
 }
 
 export interface RecaptchaConfig {
```
Comment (lahirumaramba) [inline src/auth/auth-config.ts:1551]: Please add doc strings here. Something like:
```
/**
 * The request interface for updating a ReCAPTCHA Config.
 */
```

File: etc/firebase-admin.auth.api.md [hunk 4]
```diff
@@ -264,6 +264,34 @@ export interface ProviderIdentifier {
     providerUid: string;
 }
 
+// @public
+export type RecaptchaAction = 'BLOCK';
+
+// @public (undocumented)
+export interface RecaptchaConfig {
```
Comment (lahirumaramba) [inline etc/firebase-admin.auth.api.md:271]: Please document this where you `export interface RecaptchaConfig`. API Extractor will pick it up from there. Thanks!

File: src/auth/auth-config.ts [hunk 5]
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
...
  /**
   * The enforcement state of email password provider.
   */
- emailPasswordEnforcementState?: RecaptchaProviderEnforcementState;
+  emailPasswordEnforcementState?: RecaptchaProviderEnforcementState;
```
[the remark this answers is not in the document]
Comment (Xiaoshouzi-gh) [reply src/auth/auth-config.ts:1555]: the identation of the code is correct here, but missing a space for the documentation.  Updated.

File: src/auth/auth-config.ts [hunk 6]
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
...
+
+      if (options.emailPasswordEnforcementState !== 'OFF' &&
+        options.emailPasswordEnforcementState !== 'AUDIT' &&
+        options.emailPasswordEnforcementState !== 'ENFORCE') {
+        throw new FirebaseAuthError(
+          AuthClientErrorCode.INVALID_CONFIG,
+          '"RecaptchaConfig.emailPasswordEnforcementState" must be either "OFF", "AUDIT" or "ENFORCE".',
+        );
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Xiaoshouzi-gh) [reply src/auth/auth-config.ts:1624]: The reason for this separation is to align with the backend error-code and keep consistency with the existing error message.  E.g. empty string is an INVALID_ARGUMENT but invalid string is an INVALID_CONFIG.

File: src/auth/auth-config.ts [hunk 7]
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
...
- recaptchaKeys?: RecaptchaKey[];
+  recaptchaKeys?: RecaptchaKey[];
+}
+
+export class RecaptchaAuthConfig implements RecaptchaConfig {
+  public readonly emailPasswordEnforcementState?: RecaptchaProviderEnforcementState;
+  public readonly managedRules?: RecaptchaManagedRule[];
+  public readonly recaptchaKeys?: RecaptchaKey[];
+
+  constructor(recaptchaConfig: RecaptchaConfig | undefined) {
```
[the remark this answers is not in the document]
Comment (Xiaoshouzi-gh) [reply src/auth/auth-config.ts:1573]: Remove the double undefined check from auth-config part.  Tenant request is still valid when recaptchaConfig is undefined.

File: src/auth/tenant.ts [hunk 8]
```diff
@@ -152,6 +160,9 @@ export class Tenant {
       // null will clear existing test phone numbers. Translate to empty object.
       request.testPhoneNumbers = tenantOptions.testPhoneNumbers ?? {};
     }
+    if (typeof tenantOptions.recaptchaConfig !== 'undefined') {
+      request.recaptchaConfig = tenantOptions.recaptchaConfig;
```
[the remark this answers is not in the document]
Comment (Xiaoshouzi-gh) [reply src/auth/tenant.ts:164]: Please also refer to the reply for the next comment.  Besides the logic above logic, there is also another benefit not calling `buildServerRequest` is the validation part for both `MFA` and `EmailSignInConfig` got called unnecessarily twice in this constructor.

File: src/auth/auth-config.ts [hunk 9]
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
...
+      if (!validator.isArray(options.managedRules)) {
+        throw new FirebaseAuthError(
+          AuthClientErrorCode.INVALID_CONFIG,
+          '"RecaptchaConfig.managedRules" must be an array of valid "RecaptchaManagedRule".',
+        );
+      }
+      // Validate each rule of the array
+      options.managedRules.forEach((managedRule) => {
+        ManagedRuleAuth.validate(managedRule);
+      });
```
[the remark this answers is not in the document]
Comment (Xiaoshouzi-gh) [reply src/auth/auth-config.ts:1637]: Updated.

File: src/auth/auth-config.ts [hunk 10]
```diff
@@ -1480,6 +1480,54 @@ export interface RecaptchaManagedRule {
  action?: RecaptchaAction;
 }
 
+export class ManagedRuleAuth implements RecaptchaManagedRule{
```
[the remark this answers is not in the document]
Comment (Xiaoshouzi-gh) [reply src/auth/auth-config.ts:1483]: This class has been removed. Not applicable anymore

File: src/auth/auth-config.ts [hunk 11]
```diff
@@ -1497,22 +1545,116 @@ export interface RecaptchaKey {
...
+
+  public toJSON(): object {
+    const json: any = {
+      emailPasswordEnforcementState: this.emailPasswordEnforcementState,
+      managedRules: deepCopy(this.managedRules),
+      recaptchaKeys: deepCopy(this.recaptchaKeys)
+    }
+
+    if (typeof json.emailPasswordEnforcementState === 'undefined') {
+      delete json.emailPasswordEnforcementState;
```
[the remark this answers is not in the document]
Comment (Xiaoshouzi-gh) [reply src/auth/auth-config.ts:1649]: `"emailPasswordEnforcementState": [undefined]` will be shown in the response.  This is also following the existing `multiFactorConfig` and `emailSignInConfig` class `toJSON()` method behavior.

**Comment by Xiaoshouzi-gh** [comment]
> Thank you, Liubin! I added some comments and questions. Didn't get a chance to dive into unit tests. I will take another look later. Thanks!

Thank you @lahirumaramba for the review.  Addressed majority comments.  Resolved the trivial changes and left comments on the ones needed.

File: src/auth/tenant.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (lahirumaramba) [reply src/auth/tenant.ts:167]: Good point! Thanks!

File: src/auth/tenant.ts [hunk 12]
```diff
@@ -185,6 +199,7 @@ export class Tenant {
       anonymousSignInEnabled: true,
       multiFactorConfig: true,
       testPhoneNumbers: true,
+      recaptchaConfig: true
```
Comment (lahirumaramba) [inline src/auth/tenant.ts:202]:
```suggestion
      recaptchaConfig: true,
```

**Review by lahirumaramba — APPROVED** [review]
Thank you! Just a few nits but otherwise LGTM!

**Pull request merged**